[CoordinatedGraphics] Add CoordinatedPlatformLayerBuffer to replace TextureMapperPlatformLayerBuffer

carlosgcampos opened the pull request:
#### e7b28d6a4e06fafc18cc0061368206b8a46ee1e3
<pre>
[CoordinatedGraphics] Add CoordinatedPlatformLayerBuffer to replace TextureMapperPlatformLayerBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=279052">https://bugs.webkit.org/show_bug.cgi?id=279052</a>

Reviewed by Miguel Gomez.

Add CoordinatedPlatformLayerBuffer as base class for different kind of
buffers, and add subclasses for currently supported buffers. This patch
also handles Video and HolePunch as buffer types, so that we can remove
the GStreamerVideoFrameHolder and the HolePunchClient. The plan is to
have a single TextureMapperPlatformLayerProxy with different kind of
buffers, so this patch will allow us to add a DMABuf buffer and remove
the TextureMapperPlatformLayerProxyDMABuf.

* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameHolder.cpp: Removed.
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameHolder.h: Removed.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::pushTextureToCompositor):
(WebCore::MediaPlayerPrivateGStreamer::flushCurrentBuffer):
(WebCore::MediaPlayerPrivateGStreamer::pushNextHolePunchBuffer):
(WebCore::GStreamerHolePunchClient::GStreamerHolePunchClient): Deleted.
* Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp:
(WebCore::MediaPlayerPrivateHolePunch::pushNextHolePunchBuffer):
* Source/WebCore/platform/graphics/nicosia/NicosiaPlaceholderRenderingContextSource.cpp:
(Nicosia::NicosiaPlaceholderRenderingContextSource::setPlaceholderBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLTextureMapperANGLE::platformInitialize):
* Source/WebCore/platform/graphics/texmap/TextureMapperGCGLPlatformLayer.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerBuffer.cpp: Removed.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerBuffer.h: Removed.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.cpp:
(WebCore::TextureMapperPlatformLayerProxyGL::activateOnCompositingThread):
(WebCore::TextureMapperPlatformLayerProxyGL::invalidate):
(WebCore::TextureMapperPlatformLayerProxyGL::pushNextBuffer):
(WebCore::TextureMapperPlatformLayerProxyGL::swapBuffer):
(WebCore::TextureMapperPlatformLayerProxyGL::dropCurrentBufferWhilePreservingTexture):
(WebCore::TextureMapperPlatformLayerProxyGL::scheduleUpdateOnCompositorThread):
(WebCore::TextureMapperPlatformLayerProxyGL::getAvailableBuffer): Deleted.
(WebCore::TextureMapperPlatformLayerProxyGL::appendToUnusedBuffers): Deleted.
(WebCore::TextureMapperPlatformLayerProxyGL::scheduleReleaseUnusedBuffers): Deleted.
(WebCore::TextureMapperPlatformLayerProxyGL::releaseUnusedBuffersTimerFired): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBuffer.h: Added.
(WebCore::CoordinatedPlatformLayerBuffer::type const):
(WebCore::CoordinatedPlatformLayerBuffer::size const):
(WebCore::CoordinatedPlatformLayerBuffer::flags const):
(WebCore::CoordinatedPlatformLayerBuffer::waitForContentsIfNeeded):
(WebCore::CoordinatedPlatformLayerBuffer::CoordinatedPlatformLayerBuffer):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferExternalOES.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferExternalOES::create):
(WebCore::CoordinatedPlatformLayerBufferExternalOES::CoordinatedPlatformLayerBufferExternalOES):
(WebCore::CoordinatedPlatformLayerBufferExternalOES::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferExternalOES.h: Added.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferHolePunch.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferHolePunch::create):
(WebCore::CoordinatedPlatformLayerBufferHolePunch::CoordinatedPlatformLayerBufferHolePunch):
(WebCore::m_quirksManager):
(WebCore::CoordinatedPlatformLayerBufferHolePunch::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferHolePunch.h: Added.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferRGB.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferRGB::create):
(WebCore::CoordinatedPlatformLayerBufferRGB::CoordinatedPlatformLayerBufferRGB):
(WebCore::CoordinatedPlatformLayerBufferRGB::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferRGB.h: Added.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferVideo::create):
(WebCore::CoordinatedPlatformLayerBufferVideo::CoordinatedPlatformLayerBufferVideo):
(WebCore::CoordinatedPlatformLayerBufferVideo::~CoordinatedPlatformLayerBufferVideo):
(WebCore::CoordinatedPlatformLayerBufferVideo::copyBuffer const):
(WebCore::CoordinatedPlatformLayerBufferVideo::createBufferFromGLMemory):
(WebCore::CoordinatedPlatformLayerBufferVideo::createBufferFromMainMemory):
(WebCore::CoordinatedPlatformLayerBufferVideo::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.h: Added.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferYUV.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferYUV::create):
(WebCore::CoordinatedPlatformLayerBufferYUV::CoordinatedPlatformLayerBufferYUV):
(WebCore::CoordinatedPlatformLayerBufferYUV::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferYUV.h: Added.

Canonical link: <a href="https://commits.webkit.org/283142@main">https://commits.webkit.org/283142@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f117b44534a52a075ad5af709a0c477601363bf9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65373 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44742 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17987 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69397 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15979 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67491 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52524 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16261 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11053 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68439 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41333 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56576 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/33121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38006 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13952 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14856 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14291 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71102 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9325 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13752 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59819 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9357 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56638 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/60094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7702 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1355 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9912 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40553 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41629 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42810 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41373 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp:
```diff
@@ -0,0 +1,192 @@
+
+namespace WebCore {
+
+std::unique_ptr<CoordinatedPlatformLayerBufferVideo> CoordinatedPlatformLayerBufferVideo::create(GstSample* sample, std::optional<GstVideoDecoderPlatform> videoDecoderPlatform, bool gstGLEnabled, OptionSet<TextureMapperFlags> flags)
```
perhaps use `const GRefPtr<GstSample>&` here and keep it as m_sample?

Also you could use `GstMappedFrame` also, instead of the more manual work of calling the gst_video_frame_map and friends.

carlosgcampos replied:
> perhaps use `const GRefPtr<GstSample>&` here and keep it as m_sample?

Do we need to keep a reference of the sample?

> Also you could use `GstMappedFrame` also, instead of the more manual work of calling the gst_video_frame_map and friends.

I thought about it, but decided to just try to copy existing code as much as possible in this patch.

philn replied:
> Do we need to keep a reference of the sample?

It might be safer yes. The buffer we get from it has a lifetime bound to it. Or we could also just keep a ref of the buffer.

carlosgcampos replied:
We are already keeping a ref of the buffer

magomez commented on Source/WebCore/platform/TextureMapper.cmake:
```diff
@@ -69,6 +71,13 @@ if (USE_COORDINATED_GRAPHICS)
 
+    if (USE_GSTREAMER)
+        list(APPEND WebCore_SOURCES
+            platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferHolePunch.cpp
```
This is also needed by MediaPlayerPrivateHolePunch, which doesn't require GStreamer

magomez commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp:
```diff
@@ -0,0 +1,192 @@
+    OptionSet<BitmapTexture::Flags> textureFlags;
+    if (GST_VIDEO_INFO_HAS_ALPHA(&m_videoFrame.info))
+        textureFlags.add(BitmapTexture::Flags::SupportsAlpha);
+    auto texture = BitmapTexture::create(m_size, textureFlags);
```
you're now creating a new BitmapTexture per frame, which wasn't happening before as the previous ones would be reused during the playback. What's the reason for this? Just making the code simpler? Cause this is probably going to have a performance penalty

carlosgcampos replied:
I think I forgot to move this to paintToTextureMapper like the other case. Good catch!

carlosgcampos replied:
Moved

magomez commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp:
```diff
@@ -0,0 +1,192 @@
+}
+
+std::unique_ptr<CoordinatedPlatformLayerBuffer> CoordinatedPlatformLayerBufferVideo::createBufferIfNeeded(GstVideoInfo* videoInfo, bool gstGLEnabled)
+{
```
I have a doubt here. This function is called on the constructor to initialize m_buffer. Why does it do it only for the gl case? When it's not gl, why are you waiting to paintToTextureMapper to do so? Is it with the intention of performing the gl upload during the composition?

carlosgcampos replied:
Yes, exactly, when we have a texture already we can just create the buffer that will be used directly in paintToTextureMapper. When we need to map the buffer and upload the pixels, we want to do the upload in the compositor thread, which also allows us to use the texture mapper pool.

magomez replied:
I haven't checked recently, but I think that pushTextureToCompositor is already being called in the compositor thread when using the non gstgl path (but I many be wrong and that's not the meaning of m_usingFallbackVideoSink). This would mean that you are already on the compositor thread when calling createBufferIfNeeded in the non gstgl case. Anyway, as you rearranged the code, the call to scheduleOnCompositorThread may not be needed anymore, as we wait for the actual composition to copy the buffer into a texture, unlike before.

carlosgcampos replied:
You are right, we don't want to schedule work on the compositor, I missed that, I'm going to change it.

carlosgcampos committed:
[CoordinatedGraphics] Add CoordinatedPlatformLayerBuffer to replace TextureMapperPlatformLayerBuffer
https://bugs.webkit.org/show_bug.cgi?id=279052

Reviewed by Miguel Gomez.

Add CoordinatedPlatformLayerBuffer as base class for different kind of
buffers, and add subclasses for currently supported buffers. This patch
also handles Video and HolePunch as buffer types, so that we can remove
the GStreamerVideoFrameHolder and the HolePunchClient. The plan is to
have a single TextureMapperPlatformLayerProxy with different kind of
buffers, so this patch will allow us to add a DMABuf buffer and remove
the TextureMapperPlatformLayerProxyDMABuf.

* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameHolder.cpp: Removed.
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameHolder.h: Removed.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::pushTextureToCompositor):
(WebCore::MediaPlayerPrivateGStreamer::flushCurrentBuffer):
(WebCore::MediaPlayerPrivateGStreamer::pushNextHolePunchBuffer):
(WebCore::GStreamerHolePunchClient::GStreamerHolePunchClient): Deleted.
* Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp:
(WebCore::MediaPlayerPrivateHolePunch::pushNextHolePunchBuffer):
* Source/WebCore/platform/graphics/nicosia/NicosiaPlaceholderRenderingContextSource.cpp:
(Nicosia::NicosiaPlaceholderRenderingContextSource::setPlaceholderBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::GraphicsContextGLTextureMapperANGLE::platformInitialize):
* Source/WebCore/platform/graphics/texmap/TextureMapperGCGLPlatformLayer.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerBuffer.cpp: Removed.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerBuffer.h: Removed.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.cpp:
(WebCore::TextureMapperPlatformLayerProxyGL::activateOnCompositingThread):
(WebCore::TextureMapperPlatformLayerProxyGL::invalidate):
(WebCore::TextureMapperPlatformLayerProxyGL::pushNextBuffer):
(WebCore::TextureMapperPlatformLayerProxyGL::swapBuffer):
(WebCore::TextureMapperPlatformLayerProxyGL::dropCurrentBufferWhilePreservingTexture):
(WebCore::TextureMapperPlatformLayerProxyGL::scheduleUpdateOnCompositorThread):
(WebCore::TextureMapperPlatformLayerProxyGL::getAvailableBuffer): Deleted.
(WebCore::TextureMapperPlatformLayerProxyGL::appendToUnusedBuffers): Deleted.
(WebCore::TextureMapperPlatformLayerProxyGL::scheduleReleaseUnusedBuffers): Deleted.
(WebCore::TextureMapperPlatformLayerProxyGL::releaseUnusedBuffersTimerFired): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBuffer.h: Added.
(WebCore::CoordinatedPlatformLayerBuffer::type const):
(WebCore::CoordinatedPlatformLayerBuffer::size const):
(WebCore::CoordinatedPlatformLayerBuffer::flags const):
(WebCore::CoordinatedPlatformLayerBuffer::waitForContentsIfNeeded):
(WebCore::CoordinatedPlatformLayerBuffer::CoordinatedPlatformLayerBuffer):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferExternalOES.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferExternalOES::create):
(WebCore::CoordinatedPlatformLayerBufferExternalOES::CoordinatedPlatformLayerBufferExternalOES):
(WebCore::CoordinatedPlatformLayerBufferExternalOES::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferExternalOES.h: Added.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferHolePunch.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferHolePunch::create):
(WebCore::CoordinatedPlatformLayerBufferHolePunch::CoordinatedPlatformLayerBufferHolePunch):
(WebCore::m_quirksManager):
(WebCore::CoordinatedPlatformLayerBufferHolePunch::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferHolePunch.h: Added.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferRGB.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferRGB::create):
(WebCore::CoordinatedPlatformLayerBufferRGB::CoordinatedPlatformLayerBufferRGB):
(WebCore::CoordinatedPlatformLayerBufferRGB::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferRGB.h: Added.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferVideo::create):
(WebCore::CoordinatedPlatformLayerBufferVideo::CoordinatedPlatformLayerBufferVideo):
(WebCore::CoordinatedPlatformLayerBufferVideo::~CoordinatedPlatformLayerBufferVideo):
(WebCore::CoordinatedPlatformLayerBufferVideo::copyBuffer const):
(WebCore::CoordinatedPlatformLayerBufferVideo::createBufferFromGLMemory):
(WebCore::CoordinatedPlatformLayerBufferVideo::createBufferFromMainMemory):
(WebCore::CoordinatedPlatformLayerBufferVideo::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.h: Added.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferYUV.cpp: Added.
(WebCore::CoordinatedPlatformLayerBufferYUV::create):
(WebCore::CoordinatedPlatformLayerBufferYUV::CoordinatedPlatformLayerBufferYUV):
(WebCore::CoordinatedPlatformLayerBufferYUV::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferYUV.h: Added.

Canonical link: https://commits.webkit.org/283142@main
